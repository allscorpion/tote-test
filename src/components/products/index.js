export { ProductsContainer } from './products.container';