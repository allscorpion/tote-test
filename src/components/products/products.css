.grid {
    display: grid;
    grid-gap: 24px;
    grid-template-columns: repeat(4, 1fr);
    width: 100%;
}

.grid-item {
    box-sizing: border-box;
    width: 100%;
    border-radius: 6px;
    box-shadow: 0 0 2px 3px rgba(0,0,0,0.1);
    background-color: white;
    padding: 24px;
}

.thumbnail {
    display: block;
    max-width: 100%;
}