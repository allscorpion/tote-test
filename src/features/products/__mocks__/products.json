[
    {
        "id": 1,
        "product": "Pie",
        "cost": 3.20,
        "expiryDate": "2022-08-08T00:00:00.000Z"
    },
    {
        "id": 2,
        "product": "Pie",
        "cost": 3.20,
        "expiryDate": "2022-06-08T00:00:00.000Z"
    },
    {
        "id": 3,
        "product": "Pie",
        "cost": 3.20,
        "expiryDate": "2022-08-03T00:00:00.000Z"
    },
    {
        "id": 4,
        "product": "Pie",
        "cost": 3.20,
        "expiryDate": "2022-08-12T00:00:00.000Z"
    },
    {
        "id": 5,
        "product": "Chips",
        "cost": 1.80,
        "expiryDate": null
    },
    {
        "id": 6,
        "product": "Chips",
        "cost": 1.80,
        "expiryDate": null
    },
    {
        "id": 7,
        "product": "Chips",
        "cost": 1.80,
        "expiryDate": null
    }
]