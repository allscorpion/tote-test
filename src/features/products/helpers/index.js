export * from './filter-expired-products';