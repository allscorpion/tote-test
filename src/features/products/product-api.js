import productsJson from './__mocks__/products.json';

// A mock function to mimic making an async request for data
export const getProducts = () => {
    return new Promise((resolve) => {
        return setTimeout(() => resolve({ data: productsJson }), 2000);
    });
}