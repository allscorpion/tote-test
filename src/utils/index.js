export * from './convert-currency-to-number';
export * from './discounts';
export * from './format-to-currency';
export * from './get-amount-of-days-between-dates';