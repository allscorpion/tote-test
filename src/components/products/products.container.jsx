import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { addToCart, selectCartItems } from '../../features/cart';
import { fetchProducts, selectProducts } from '../../features/products/product-slice';
import { Products } from './products';

export const ProductsContainer = () => {

    const productItems = useSelector(selectProducts);
    const cartTtems = useSelector(selectCartItems);
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(fetchProducts());
    }, [dispatch]);

    const addItemToBasket = (item) => {
        dispatch(addToCart(item));
    };

    if (!productItems.length) {
        return <p>Loading...</p>;
    } 

    return (
        <Products 
            items={productItems} 
            cartItems={cartTtems}
            addItemToBasket={addItemToBasket} 
        />
    );
};