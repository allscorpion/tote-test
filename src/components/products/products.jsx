import React from 'react';
import pie from '../../images/pie.jpg';
import chips from '../../images/chips.jpg';
import './products.css';

export const Products = (props) => {
    const {items, addItemToBasket, cartItems} = props;
    return (
        <div className="grid">
            {items.map(item => {
                const itemIsInCart = cartItems?.some(x => x.id === item.id);
                return (
                    <div className='grid-item' key={item.id}>
                        {/* thumbnail would usually come from the api */}
                        {item.product.toLowerCase() === 'pie' && <img src={pie} className="thumbnail" alt="Pie" />}
                        {item.product.toLowerCase() === 'chips' && <img src={chips} className="thumbnail" alt="Chips" />}
                        <h2>{item.product}</h2>
                        <p>£{item.cost}</p>
                        {item.discountPrice !== undefined && <p>Discount price - £{item.discountPrice}</p>}
                        <button 
                            onClick={() => {
                                addItemToBasket(item);
                            }}
                            disabled={itemIsInCart}
                        >
                            {!itemIsInCart ? 'Add' : 'Added'}
                        </button>
                    </div>
                );
            })}
        </div>
    );
}