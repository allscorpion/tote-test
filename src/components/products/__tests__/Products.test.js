import React from 'react';
import { renderWithProviders } from '../../../../__tests__/utils/test-utils';
import { Products } from '../products';


describe('Product tests', () => {
    it('renders the loader when there\'s no items', () => {
        const render = renderWithProviders(<Products items={[]} />);

        expect(render.asFragment()).toMatchSnapshot();
    });

    it('renders the items when we have some', () => {
        const render = renderWithProviders(<Products items={[
            {
                "id": 1,
                "product": "Pie",
                "cost": 3.20,
                "expiryDate": "2022-08-08T00:00:00.000Z"
            }
        ]} />);

        expect(render.asFragment()).toMatchSnapshot();
    });
})