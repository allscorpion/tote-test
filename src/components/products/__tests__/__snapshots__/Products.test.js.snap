// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product tests renders the items when we have some 1`] = `
<DocumentFragment>
  <div
    class="grid"
  >
    <div
      class="grid-item"
    >
      <img
        alt="Pie"
        class="thumbnail"
        src="pie.jpg"
      />
      <h2>
        Pie
      </h2>
      <p>
        Â£3.2
      </p>
      <button>
        Add
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Product tests renders the loader when there's no items 1`] = `
<DocumentFragment>
  <div
    class="grid"
  />
</DocumentFragment>
`;
